Add pids-limit support in docker update

darkowlzz opened the pull request:
- Added a new flag `--pids-limit` in `docker update` command.
- Value of the `pidsLimit` flag variable is set as `PidsLimit` in
`containertypes.Resources`, which is used to apply the update in
hostConfig and the real world (running containers).
- Added `TestUpdatePIDsLimit` integration test for the new flag.

Fixes #32443

Signed-off-by: Sunny Gogoi <indiasuny000@gmail.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Added a new flag, `--pids-limit` , in the command `docker update`.

**- How I did it**

Details are in commit body.

**- How to verify it**

Verify by starting a container with `--pids-limit=4` and then updating pids-limit of the container
to a new value with `docker update --pids-limit=<newValue> <containerName>`.
Once updated, the result can be seen in `docker inspect <containerName>` and in the file
`/sys/fs/cgroup/pids/pids.max` within the container.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

Add pids-limit support in docker update

**- A picture of a cute animal (not mandatory but encouraged)**

![cute animal](https://s-media-cache-ak0.pinimg.com/736x/fd/48/67/fd48674fda8c9054ec1a766a5d896b9e--no-drama-baby-llama.jpg)

darkowlzz committed:
Add pids-limit support in docker update

- Adds updating PidsLimit in UpdateContainer().
- Adds setting PidsLimit in toContainerResources().

Signed-off-by: Sunny Gogoi <indiasuny000@gmail.com>
Signed-off-by: Brian Goff <cpuguy83@gmail.com>

darkowlzz commented on daemon/update_linux.go:
```diff
@@ -28,5 +28,6 @@ func toContainerdResources(resources container.Resources) libcontainerd.Resource
 	}
 	r.MemoryReservation = uint64(resources.MemoryReservation)
 	r.KernelMemoryLimit = uint64(resources.KernelMemory)
+	r.PidsLimit = uint64(resources.PidsLimit)
```
Need help with this part. From my findings, I need to add `PidsLimit` here to update the running containers (real world). But `libcontainered.Resources` has no attribute named `PidsLimit`. Hence, the build would fail for now.

I found that the libcontainered resources attributes are defined in (github.com/docker/containerd/api/grpc/types)[https://github.com/docker/docker/blob/master/vendor/github.com/docker/containerd/api/grpc/types/api.proto], but I'm not sure how to generate `api-pb.go` once I edit `api.proto`.

Can someone help me with this?

thaJeztah replied:
I think @mlaventure would be the best person to assist 👍

darkowlzz replied:
So I read proto3 docs and installed `protoc` with Go protocol buffer plugin.
Ran `protoc --go_out=./ api.proto` from `vendor/github.com/docker/containerd/api/grpc/types`. It generated `api.pb.go` with `PidsLimit` in `UpdateResource` and a method `GetPidsLimit()` for the same type. Along with this, it also modified a lot of other parts of the file. Not sure if we need so many changes.

I'll wait for an assistance on how to generate `api.pb.go` that fits here 🙂

mlaventure replied:
@darkowlzz you need to open a PR against the `v0.2.x` branch of `https://github.com/containerd/containerd` to add that field into `api.proto`. Once that PR is merged you'll need to vendor `containerd` into docker (as part of this PR)

thaJeztah commented on cli/command/container/update.go:
```diff
@@ -66,6 +67,7 @@ func NewUpdateCommand(dockerCli *command.DockerCli) *cobra.Command {
 	flags.Var(&opts.memorySwap, "memory-swap", "Swap limit equal to memory plus swap: '-1' to enable unlimited swap")
 	flags.Var(&opts.kernelMemory, "kernel-memory", "Kernel memory limit")
 	flags.StringVar(&opts.restartPolicy, "restart", "", "Restart policy to apply when a container exits")
+	flags.Int64Var(&opts.pidsLimit, "pids-limit", 0, "Tune container pids limit (set -1 for unlimited)")
```
Before I forget to comment; this needs an "annotation" so that the flag is hidden when connecting to an older daemon version;

```go
flags.SetAnnotation("cpus", "version", []string{"1.30"})
```

darkowlzz replied:
Added "annotation" for "pids-limit" with version 1.30.

@thaJeztah do you want me to change annotation version for "cpus"? Not sure why your code snippet has "cpus" with version 1.30, when the same line exists with 1.29 😅

thaJeztah replied:
oh, sorry, no that was just a copy/paste error on my side

mlaventure replied:
btw: Shouldn't the default be `-1` and a value of `0` errors out?

darkowlzz commented:
After vendoring new gRPC APIs, build is no longer failing but when I update pids-limit, HostConfig gets updated with new limit but not the file `/sys/fs/cgroup/pids/pids.max` in the container. I couldn't figure out what am I missing. Till [here](https://github.com/docker/docker/blob/master/libcontainerd/client_linux.go#L290) `PidsLimit` is passed with the new value. And then nothing happens. `/sys/fs/cgroup/pids/pids.max` doesn't change.

@thaJeztah & @mlaventure can you guys help here? Couldn't find exactly where cgroup files are updated.

mlaventure commented:
@darkowlzz the path would be something like `/sys/fs/cgroup/pids/docker/<container-id>/pids.max`

Also, you need to update `vendor.conf` with the last `containerd` commit and run `vndr`

darkowlzz commented:
@mlaventure oops! 😅 fixed the vendoring.

And about the pids-limit file, yes, `/sys/fs/cgroup/pids/docker/<container-id>/pids.max` is the file on host machine, that reflects `/sys/fs/cgroup/pids/pids.max` in the container. But I'm stuck with `pids.max` file not getting updated even after passing `PidsLimit` to [UpdateContainer()](https://github.com/docker/docker/blob/master/libcontainerd/client_linux.go#L290).

On checked runc code (rev 9c2d8d184e5da67c95d601382adf14862e4f2228, from `vendor.conf`), I found that [runc/libcontainer/cgroups/fs/pids.go](https://github.com/docker/runc/blob/9c2d8d184e5da67c95d601382adf14862e4f2228/libcontainer/cgroups/fs/pids.go#L38) has the code that updates `pids.max`. But for some reason, that code is not in the vendored runc in docker repo. But at the same time `runc/libcontainer/cgroup/fs/blkio.go` and other similar files are also not vendored but their corresponding files do get updated 🤷‍♂️ confused. Need help in understanding how it's working  😞

mlaventure reviewed (commented):
@darkowlzz sorry for the delay, DockerCon happened :)

I think it doesn't work for you because you forgot to also update `hack/dockerfile/binaries-commits` :)

The `runc` code for pids doesn't get vendored in because it is not directly used by the engine code (there's no reference to either of the type or routine found in that package within the engine). The use happen indirectly via `containerd`

darkowlzz commented:
@mlaventure Changed the default pidsLimit value to -1 for max value by default 😅

Updated `hack/dockerfile/binaries-commits` with new containerd commit hash. To install new containerd, ran `hack/dockerfile/install-binaries.sh containerd`. This failed due to recent changes in containerd import path `docker/containerd` -> `containerd/containerd`.  Changed `install-binaries.sh` to fix this and it installed the new containerd successfully.

Checked vendored `containerd/api/grpc/types/api.pb.go` to ensure the new code is present.

Rebuilt docker `hack/make.sh binary install-binary run`.

Checked containerd version in `docker info`
`containerd version: a47bf53545fed63fe49646b289c36b57c5386b06`

Created a new alpine container with --pids-limit set to 20 and then updated to 15. Updated value is visible in `docker inspect`. But it still doesn't update in `/sys/fs/cgroup/pids/pids.max` or `/sys/fs/cgroup/pids/docker/<container-id>/pids.max`.

Anything else that is missing?

mlaventure commented:
@darkowlzz everything is now fine on both `docker` and `containerd`. It looks like `runc` needs to be updated though :/

I made a PR to add it [here](https://github.com/opencontainers/runc/pull/1423)

darkowlzz commented:
@mlaventure 😱 great! thanks a lot :)

darkowlzz commented:
yay! runc changes got merged 🎉 @mlaventure 

In `vendor.conf`, I see a fork of runc (docker/runc) is being used. Would be it safe to update that with the main opencontainers/runc repo?

mlaventure commented:
@darkowlzz yes, we are just waiting for `runc` master to be merged into `containerd` see https://github.com/containerd/containerd/pull/778

It's waiting on the integration with `docker` to be done and tested.

cpuguy83 commented:
Looks like we need a rebase, and move the CLI bits to github.com/docker/cli now that it has been moved out of this repo.

darkowlzz commented:
Rebased again. Moved the docs changes to docker/cli PR.

Should I remove `TestUpdatePIDsLimit` from cli integration test? I'm not sure how would we test it if this PR should land before docker/cli PR.

cpuguy83 commented:
@darkowlzz Can you convert the test to only use the API?

darkowlzz commented:
It works! 😃

cpuguy83 reviewed (approved):
Design LGTM

cpuguy83 commented:
Maybe we can add a test to verify that indeed the container is now respecting the new limit?

darkowlzz commented:
> Maybe we can add a test to verify that indeed the container is now respecting the new limit?

@cpuguy83 need some help with that. In the test, I start a container in daemon mode with PidsLimit set to 5 and top running. Then I run `docker exec` with `top &` multiple times to create enough number of processes. But on reaching the PID limit, exec fails with error:
```
rpc error: code = Unknown desc = oci runtime error: exec failed: container_linux.go:262: starting container process caused "read init-p: connection reset by peer"
```

Even tried `sh -c top & top & top & ....` in a single exec, but fails with the same error.

Is there any other better way to run parallel processes? Maybe something really simple that I'm missing 🤔

cpuguy83 commented:
@darkowlzz That's an interesting error. Seems like that's a bug, at least in the error reporting but also like we aren't consuming everything in a pipe.

For the test you can setup multiple commands that stay attached and expect the last one to error.

darkowlzz commented on integration-cli/docker_api_update_unix_test.go:
```diff
@@ -34,3 +34,40 @@ func (s *DockerSuite) TestAPIUpdateContainer(c *check.C) {
+	dockerCmd(c, "exec", name, "top")
+
+	_, _, err := dockerCmdWithError("exec", name, "top")
+	c.Assert(err, checker.NotNil)
```
I'm not sure what's wrong here. Whenever I use `dockerCmdWithError`, the test gets stuck and I have to stop it or it times out after a long time. idk why it's not working here, when other tests are using this in the same way :(

cpuguy83 commented:
Because it's staying attached. You'd have to use `-d`, but then you also
need to make sure each `top` is running prior before running your check.

On Wed, Jun 21, 2017 at 5:16 PM, Sunny <notifications@github.com> wrote:

> *@darkowlzz* commented on this pull request.
> ------------------------------
>
> In integration-cli/docker_api_update_unix_test.go
> <https://github.com/moby/moby/pull/32519#discussion_r123396082>:
>
> > +
> +func (s *DockerSuite) TestAPIContainerPidsLimit(c *check.C) {
> +	testRequires(c, DaemonIsLinux)
> +	testRequires(c, pidsLimit)
> +
> +	name := "apiContainerPidsLimit"
> +	dockerCmd(c, "run", "-d", "--name", name, "--pids-limit", "5", "busybox", "top")
> +	c.Assert(inspectField(c, name, "HostConfig.PidsLimit"), checker.Equals, "5")
> +
> +	dockerCmd(c, "exec", name, "top")
> +	dockerCmd(c, "exec", name, "top")
> +	dockerCmd(c, "exec", name, "top")
> +	dockerCmd(c, "exec", name, "top")
> +
> +	_, _, err := dockerCmdWithError("exec", name, "top")
> +	c.Assert(err, checker.NotNil)
>
> I'm not sure what's wrong here. Whenever I use dockerCmdWithError, the
> test gets stuck and I have to stop it or it times out after a long time.
> idk why it's not working here, when other tests are using this in the same
> way :(
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/moby/moby/pull/32519#pullrequestreview-45595052>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAwxZtOOSedmtya_I-0fl9gAQpE0UF6pks5sGbJegaJpZM4M5_XZ>
> .
>



-- 


- Brian Goff

darkowlzz commented:
@cpuguy83 thanks. Detaching helped in avoiding the previous problem. Now, I'm getting the connection issue unpredictable.

```
Error response from daemon: rpc error: code = Unknown desc = oci runtime error: exec failed: container_linux.go:262: starting container process caused "read init-p: connection reset by peer"
```

Sometimes, running 3 execs result in this error, but sometimes, just running 2 execs results in the same error.

cpuguy83 commented:
I've rebased this PR and moved the tests into integration/.

The one thing remaining is being able to update a pids limit back to unlimited.
This is the same issue we have with memory updates.
I do have it implemented but the API is a little weird in that the user must send `-1` on the API and this gets stored as 0... with this we always send the pids limit update to containerd.
As an alternative, we could change the pid limit in HostConfig to a pointer....

cyphar commented:
@cpuguy83 

> I do have it implemented but the API is a little weird in that the user must send -1 on the API and this gets stored as 0...

My apologies for making the `pids.max` API like that. It was an off-by-one originally 😢. Semantically speaking, we are actually writing `max` and the numeric value is sort of meaningless.

cpuguy83 commented:
No worries, it's the whole update API in docker that's tricky in that it merges instead of replaces... or doesn't deal with the zero value well.

I'm thinking we should just make the integer values pointers, but I'm sure people will hate this change when they update their vendored types.

vdemeester commented:
ping @cpuguy83 needs a rebase again 😓

thaJeztah commented:
ping @cpuguy83 😇

vdemeester commented:
@cpuguy83 @thaJeztah what's the status for this one ?

cpuguy83 commented:
Do we have some idea how to handle this? Should we change the API type to a pointer so we can differentiate the zero value (unlimited) from the unset value?

thaJeztah commented:
Sounds like the best option; not sure if we can change that for other properties that have the same issue though 😞

olljanat commented:
@darkowlzz this one needs to be rebased

olljanat commented:
Derek add label: status/failing-ci

cpuguy83 commented:
Rebased, made pids limit a pointer so users can update to unlimited by setting 0.

thaJeztah reviewed (approved):
LGTM

kolyshkin commented on integration/container/update_linux_test.go:
```diff
@@ -101,3 +104,85 @@ func TestUpdateCPUQuota(t *testing.T) {
+				assert.Equal(t, *inspect.HostConfig.Resources.PidsLimit, test.expect)
+			}
+
+			exec, err := client.ContainerExecCreate(ctx, cID, types.ExecConfig{
```
All this code (ContainerExecCreate/Attach/Start/StdCopy) can be replaced with a call to `container.Exec()`. 
Something like:
```go
res, err := container.Exec(ctx, client, resp.ID, []string{"cat", "/sys/fs/cgroup/pids/pids.max"})
assert.NilError(t, err)
assert.Assert(t, is.Len(res.Stderr(), 0))
out := res.Stdout()
```

cpuguy83 replied:
Updated.

kolyshkin reviewed (changes requested):
test case can be simlplified a lot wrt exec

thaJeztah commented:
experimental failing on a flaky test  https://github.com/moby/moby/issues/38521

```
23:06:11 FAIL: docker_cli_start_test.go:189: DockerSuite.TestStartReturnCorrectExitCode
23:06:11 
23:06:11 docker_cli_start_test.go:194:
23:06:11     c.Assert(err, checker.NotNil)
23:06:11 ... value = nil
```

powerpc failing on another flaky https://github.com/moby/moby/issues/38720

```
23:45:34 FAIL: docker_cli_daemon_test.go:118: DockerDaemonSuite.TestDaemonRestartUnlessStopped
23:45:34 
23:45:34 [dc3c44c677a52] waiting for daemon to start
23:45:34 [dc3c44c677a52] daemon started
23:45:34 
23:45:34 docker_cli_daemon_test.go:157:
23:45:34     // both stopped
23:45:34     testRun(map[string]bool{"top1": false, "top2": false, "exit": false}, "")
23:45:34 docker_cli_daemon_test.go:140:
23:45:34     c.Assert(strings.Contains(out, name), check.Equals, shouldRun, check.Commentf(format, prefix, name))
23:45:34 ... obtained bool = true
23:45:34 ... expected bool = false
23:45:34 ... container "exit" is running
23:45:34 
23:45:34 [dc3c44c677a52] exiting daemon
23:45:37 
```


I'll merge
